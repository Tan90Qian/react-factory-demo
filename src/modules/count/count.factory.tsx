import React from "react";
import { observer } from "mobx-react";

import { Injectable } from "src/core/inject";
import { loading } from "src/models/loading";
import { BaseFactory } from "src/core/baseFactory";

import { CountStore, CountPresenter } from "./count.model";
import { CountView } from "./views/count.view";
import { GlobalStores, Services } from "src/global";

@Injectable()
export class CountFactory extends BaseFactory<GlobalStores, Services> {
  constructor(private countStore: CountStore) {
    super();
    this.didMount = this.didMount.bind(this);
  }

  onClick() {
    CountPresenter.setCount(this.countStore, this.countStore.count + 1);
  }

  @loading()
  async didMount() {
    const newCount:number = await new Promise((res, rej) => {
      setTimeout(() => {
        res(10);
      }, 2000);
    });
    CountPresenter.setCount(this.countStore, newCount);
  }

  create = () => {
    return observer(() => (
      <CountView
        loading={
          this.globalStores.loadingStore.actions["CountFactory/didMount"]
        }
        count={this.countStore.count}
        onClick={this.onClick}
        didMount={this.didMount}
      />
    ));
  };
}
