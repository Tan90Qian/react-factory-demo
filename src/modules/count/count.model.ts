import { observable, action, runInAction } from "mobx";

import { BaseService } from "src/services";
import { Injectable } from "src/core/inject";

export class CountStore {
  @observable count: number;

  constructor(initialCount?: number) {
    runInAction(() => {
      this.count = initialCount || 1;
    });
  }
}

@Injectable()
export class CountPresenter {
  constructor(private service: BaseService) {}

  @action
  setCount(store: CountStore, count: number) {
    store.count = count;
  }

  @action
  async fetchCount() {
    return this.service.fetch()
  }

  @action
  static async fetchCount(service: BaseService) {
    return service.fetch()
  }
}

export class LoadingStore {
  @observable loading: boolean = false;
}

export class LoadingPresenter {
  @action
  static setLoading(store: LoadingStore, loading: boolean) {
    store.loading = loading;
  }
}
