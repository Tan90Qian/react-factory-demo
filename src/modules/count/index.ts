import { CountFactory } from "./count.factory";

export default CountFactory;
