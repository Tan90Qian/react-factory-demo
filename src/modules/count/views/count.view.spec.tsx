import React from "react";
import { shallow } from "enzyme";
import { renderHook } from "@testing-library/react-hooks";

import { CountUI, CountView, useDidMount } from "./count.view";

describe("test CountUI", () => {
  test("snapshot test", () => {
    const wrapper = shallow(
      <CountUI count={1} loading={false} onClick={() => {}} />
    );
    expect(wrapper.getElement()).toMatchSnapshot();
  });

  test("snapshot test when loading", () => {
    const wrapper = shallow(
      <CountUI count={1} loading={true} onClick={() => {}} />
    );
    expect(wrapper.getElement()).toMatchSnapshot();
  });
});

test("test useDidMount", () => {
  const didMount = jest.fn();
  renderHook(() => useDidMount(didMount));
  expect(didMount.mock.calls.length).toBe(1);
});

describe("test CountView", () => {
  const didMount = jest.fn();

  test("snapshot test", () => {
    const wrapper = shallow(
      <CountView
        didMount={didMount}
        count={1}
        loading={false}
        onClick={() => {}}
      />
    );
    expect(wrapper.getElement()).toMatchSnapshot();
  });
});
