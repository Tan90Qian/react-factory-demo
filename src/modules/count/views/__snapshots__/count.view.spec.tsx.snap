// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test CountUI snapshot test 1`] = `
<div>
  <p>
    count: 
    1
  </p>
  <button
    onClick={[Function]}
  >
    click
  </button>
</div>
`;

exports[`test CountUI snapshot test when loading 1`] = `
<p>
  loading...
</p>
`;

exports[`test CountView snapshot test 1`] = `
<CountUI
  count={1}
  loading={false}
  onClick={[Function]}
/>
`;
